src/vs/editor/browser/services/treeSitter/treeSitterParserService.ts

```diff
@@ -0,0 +1,347 @@
+	get tree() { return this._treeSitterTree; }
+
+	constructor(readonly model: ITextModel,
+		private readonly _treeSitterParser: TreeSitterLanguages,
+		private readonly _treeSitterImporter: TreeSitterImporter,
+		private readonly _logService: ILogService,
+		private readonly _telemetryService: ITelemetryService
+	) {
+		super();
+		this._register(Event.runAndSubscribe(this.model.onDidChangeLanguage, (e => this._onDidChangeLanguage(e ? e.newLanguage : this.model.getLanguageId()))));
+	}
+
+	private readonly _languageSessionDisposables = this._register(new DisposableStore());
+	/**
+	 * Be very careful when making changes to this method as it is easy to introduce race conditions.
+	 */
+	private async _onDidChangeLanguage(languageId: string) {
+		this._languageSessionDisposables.clear();
+		this._treeSitterTree = undefined;
+
+		const token = cancelOnDispose(this._languageSessionDisposables);
+		const language = await this._treeSitterParser.getLanguage(languageId);
+		if (!language || token.isCancellationRequested) {
+			return;
+		}
+
+		const Parser = await this._treeSitterImporter.getParserClass();
+		if (token.isCancellationRequested) {
+			return;
+		}
+
+		const treeSitterTree = this._languageSessionDisposables.add(new TreeSitterTree(new Parser(), language, this._logService, this._telemetryService));
+		this._languageSessionDisposables.add(this.model.onDidChangeContent(e => this._onDidChangeContent(treeSitterTree, e)));
+		await this._onDidChangeContent(treeSitterTree);
+		if (token.isCancellationRequested) {
+			return;
+		}
+
+		this._treeSitterTree = treeSitterTree;
+	}
+
+	private async _onDidChangeContent(treeSitterTree: TreeSitterTree, e?: IModelContentChangedEvent) {
+		return treeSitterTree.onDidChangeContent(this.model, e);
+	}
+}
+
+export class TreeSitterTree implements IDisposable {
+	private _tree: Parser.Tree | undefined;
+	private _isDisposed: boolean = false;
+	constructor(public readonly parser: Parser,
+		public /** exposed for tests **/ readonly language: Parser.Language,
+		private readonly _logService: ILogService,
+		private readonly _telemetryService: ITelemetryService) {
+		this.parser.setTimeoutMicros(50 * 1000); // 50 ms
+		this.parser.setLanguage(language);
+	}
+	dispose(): void {
+		this._isDisposed = true;
+		this._tree?.delete();
+		this.parser?.delete();
+	}
+	get tree() { return this._tree; }
+	set tree(newTree: Parser.Tree | undefined) {
+		this._tree?.delete();
+		this._tree = newTree;
+	}
+	get isDisposed() { return this._isDisposed; }
+
+	private _onDidChangeContentQueue: Promise<void> = Promise.resolve();
+	public async onDidChangeContent(model: ITextModel, e?: IModelContentChangedEvent) {
+		this._onDidChangeContentQueue = this._onDidChangeContentQueue.then(() => {
+			if (this.isDisposed) {
+				// No need to continue the queue if we are disposed
+				return;
+			}
+			return this._onDidChangeContent(model, e);
+		}).catch((e) => {
+			this._logService.error('Error parsing tree-sitter tree', e);
```

Comment:
I think `onUnexpectedError` might also be good here, because then you'd see it in telemetry.